### PR TITLE
feat: API org user scope

### PR DESCRIPTION
**Pull request by joeauyeung**
## What does this PR do?

This PR adds organization user scopes for V2 API. Merge this PR instead of #15468

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A - I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

yarn test:e2e organizations-users.e2e-spec.ts

- Make a GET call to `organization/{orgId}/users` to get all org users
- Make a GET call to `organization/{orgId}/users` passing a single email to get that single org user
- Make a GET call to `organization/{orgId}/users` passing an array of emails to get a filtered list of org users
- Make a POST call to `organization/{orgId}/users` passing an email to create an org user
	- Ensure that the signup to org email is sent
- Make a POST call to `organization/{orgId}/users` passing an email  and non-default user values to create an org user
	- Ensure that the user return has the updated values
- Make a PATCH call to `organization/{orgId}/users/{userId}` updating a user's data
- Make a DELETE call to `organization/{orgId}/users/{userId}` to delete the user

### CODE REVIEW

File: apps/api/v2/src/modules/organizations/controllers/users/organizations-users.e2e-spec.ts
Comment (joeauyeung) [inline apps/api/v2/src/modules/organizations/controllers/users/organizations-users.e2e-spec.ts:1]: Tests are are a WIP

File: apps/api/v2/src/modules/organizations/controllers/users/organizations-users.e2e-spec.ts [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+    });
+
+    // it("should create a new org user", async () => {
+    //   const newOrgUser = {
+    //     email: "new-member@org.com",
+    //     organizationRole: "MEMBER",
+    //     autoAccept: true,
+    //   }
+
+    //   // const emailSpy = jest.spyOn(EmailService, "sendSignupToOrganizationEmail")
```
Comment (joeauyeung) [inline apps/api/v2/src/modules/organizations/controllers/users/organizations-users.e2e-spec.ts:253]: What's the best way to spy and mock this method? I want to ensure it's being called with the right values

File: apps/api/v2/src/modules/organizations/inputs/create-organization-user.input.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { CreateUserInput } from "@/modules/users/inputs/create-user.input";
+import { MembershipRole } from "@prisma/client";
+import { IsString, IsOptional, IsBoolean, IsEnum } from "class-validator";
+
+export class CreateOrganizationUserInput extends CreateUserInput {
```
Comment (joeauyeung) [inline apps/api/v2/src/modules/organizations/inputs/create-organization-user.input.ts:5]: I created a base `User` DTO that we can then expand on for org users. The base `CreateUserInput` was taken from V1 API `apps/api/v1/lib/validations/user.ts`

File: apps/api/v2/src/modules/organizations/inputs/create-organization-user.input.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  @IsString()
+  locale = "en";
+
+  @IsOptional()
+  @IsEnum(MembershipRole)
+  organizationRole: MembershipRole = MembershipRole.MEMBER;
+
+  @IsOptional()
+  @IsBoolean()
+  autoAccept = true;
```
Comment (joeauyeung) [inline apps/api/v2/src/modules/organizations/inputs/create-organization-user.input.ts:16]: Assume that if we are creating org users via API we want the user to be automatically accepted into the org.

File: apps/api/v2/src/modules/organizations/repositories/organizations-users.repository.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    return {
+      OR: [
+        { organizationId: orgId },
+        {
+          teams: {
+            some: {
+              teamId: orgId,
+            },
+          },
+        },
```
Comment (joeauyeung) [inline apps/api/v2/src/modules/organizations/repositories/organizations-users.repository.ts:23]: Added for any backwards compatibility

File: apps/api/v2/src/modules/organizations/services/organizations-users-service.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+    return users;
+  }
+
+  async createUser(org: Team, userCreateBody: CreateOrganizationUserInput) {
+    // Check if email exists in the system
+    const userEmailCheck = await this.organizationsUsersRepository.getOrganizationUserByEmail(
+      org.id,
+      userCreateBody.email
+    );
```
Comment (joeauyeung) [inline apps/api/v2/src/modules/organizations/services/organizations-users-service.ts:36]: If there is an email conflict do we need to throw a custom error message or do we pass the Prisma error?

File: apps/api/v2/src/modules/organizations/services/organizations-users-service.ts [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+    if (userEmailCheck) throw new ConflictException("A user already exists with that email");
+
+    // Check if username is already in use in the org
+    if (userCreateBody.username) {
+      await this.checkForUsernameConflicts(org.id, userCreateBody.username);
```
Comment (joeauyeung) [inline apps/api/v2/src/modules/organizations/services/organizations-users-service.ts:42]: The same username can exist in the system and within an org scope.

File: apps/api/v2/src/modules/organizations/services/organizations-users-service.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+          orgId: org.id,
+          autoAccept: userCreateBody.autoAccept,
+        },
+      },
+    });
+
+    const createdUser = createdUserCall[0];
+
+    // Update user fields that weren't included in createNewUsersConnectToOrgIfExists
+    const updateUserBody = plainToInstance(CreateUserInput, userCreateBody, { strategy: "excludeAll" });
```
Comment (joeauyeung) [inline apps/api/v2/src/modules/organizations/services/organizations-users-service.ts:70]: Filter out any org specific fields from the update user field

File: apps/api/v2/src/modules/users/validators/localeValidator.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
+import type { ValidatorConstraintInterface } from "class-validator";
+import { ValidatorConstraint } from "class-validator";
+
+@ValidatorConstraint({ name: "localeValidator", async: false })
+export class LocaleValidator implements ValidatorConstraintInterface {
+  validate(locale: string) {
+    const localeValues = [
```
Comment (joeauyeung) [inline apps/api/v2/src/modules/users/validators/localeValidator.ts:7]: Taken from `apps/api/v1/lib/validations/user.ts`

File: apps/api/v2/src/modules/users/outputs/get-users.output.ts
Comment (joeauyeung) [inline apps/api/v2/src/modules/users/outputs/get-users.output.ts:1]: Taken from `apps/api/v1/lib/validations/user.ts`

File: apps/api/v2/src/modules/users/outputs/get-users.output.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
+import { Type } from "class-transformer";
+import { Expose } from "class-transformer";
+import { IsBoolean, IsDateString, IsInt, IsString, ValidateNested, IsArray } from "class-validator";
+
+export class GetUserOutput {
+  @IsInt()
+  @Expose()
+  id!: number;
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/users/outputs/get-users.output.ts:8]: when inputs/outputs class are defined outside of api-v2 nestjs cli is not able to auto generate the docs that's why we need to add @ApiProperty

File: apps/api/v2/src/modules/organizations/services/organizations-users-service.ts [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+    // Update new user with other userCreateBody params
+    const user = await this.organizationsUsersRepository.updateOrganizationUser(
+      org.id,
+      createdUser.id,
+      updateUserBody
+    );
+
+    // Need to send email to new user to create password
+    await this.emailService.sendSignupToOrganizationEmail({
```
Comment (joeauyeung) [inline apps/api/v2/src/modules/organizations/services/organizations-users-service.ts:80]: I should add we need to send an email to the created user so they can setup their password.

File: apps/api/v2/src/modules/organizations/services/organizations-users-service.ts [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    private readonly emailService: EmailService
+  ) {}
+
+  async getUsers(orgId: number, emailInput?: string | string[]) {
+    const emailArray = !emailInput ? [] : Array.isArray(emailInput) ? emailInput : [emailInput];
+
+    const users = await this.organizationsUsersRepository.getOrganizationUsers(orgId, emailArray);
+
+    return users;
+  }
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/services/organizations-users-service.ts:29]: I'd prefer if we didn't have to deal with inputs format outside of class validator, this is where we should already transform a string email to an array of one email if necessary

File: apps/api/v2/src/modules/users/inputs/get-users.input.ts [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+import { IsEmailStringOrArray } from "../validators/isEmailStringOrArray";
+
+export class GetUsersInput {
+  @IsOptional()
+  @Validate(IsEmailStringOrArray)
+  @ApiProperty({
+    description: "The email address or an array of email addresses to filter by",
+  })
+  email?: string | string[];
+}
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/users/inputs/get-users.input.ts:13]: here we should transform the value to an array if it is a string

File: apps/api/v2/src/modules/organizations/repositories/organizations-users.repository.ts [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+      ],
+    };
+  }
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/repositories/organizations-users.repository.ts:26]: I think we should only filter on profile.organizationId

File: apps/api/v2/src/modules/users/inputs/get-users.input.ts (hunk 12 above)
Comment (ThyMinimalDev) [reply apps/api/v2/src/modules/users/inputs/get-users.input.ts:13]: @Transform(({ value }: { value: string | string[] }) => {
    return typeof value === "string" ? [value] : value;
  })
email?: string[];

File: apps/api/v2/src/modules/organizations/controllers/users/organizations-users.e2e-spec.ts [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+        console.log("2");
+        expect(emailSpy).toHaveBeenCalledWith({
+          usernameOrEmail: newOrgUser.email,
+          orgName: org.name,
+          orgId: org.id,
+          locale: undefined,
+        });
+        console.log("3");
+        expect(false).toBe(true);
+        console.log("4");
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/controllers/users/organizations-users.e2e-spec.ts:267]: console logs

File: apps/api/v2/src/modules/organizations/controllers/users/organizations-users.e2e-spec.ts [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+        expect(userData.email).toBe(newOrgUser.email);
+        console.log("2");
+        expect(emailSpy).toHaveBeenCalledWith({
+          usernameOrEmail: newOrgUser.email,
+          orgName: org.name,
+          orgId: org.id,
+          locale: undefined,
+        });
+        console.log("3");
+        expect(false).toBe(true);
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/controllers/users/organizations-users.e2e-spec.ts:266]: 🎗️

File: apps/api/v2/src/modules/organizations/controllers/users/organizations-users.e2e-spec.ts [hunk 16]
```diff
@@ -0,0 +1,282 @@
...
+    it("should update an org user", async () => {
+      const { body } = await request(app.getHttpServer())
+        .patch(`/v2/organizations/${org.id}/users/${user.id}`)
+        .send({
+          theme: "light",
+        })
+        .set("Content-Type", "application/json")
+        .set("Accept", "application/json");
+
+      const userData = body.data;
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/controllers/users/organizations-users.e2e-spec.ts:232]: no tests here, should test if data has been modified

File: apps/api/v2/src/modules/organizations/controllers/users/organizations-users.controller.ts [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+  @Roles("ORG_ADMIN")
+  async deleteOrganizationUser(
+    @Param("orgId", ParseIntPipe) orgId: number,
+    @Param("userId", ParseIntPipe) userId: number,
+    @GetOrg() org: Team
+  ): Promise<ApiResponse<string>> {
+    const user = await this.organizationsUsersService.deleteUser(orgId, userId);
+    return {
+      status: SUCCESS_STATUS,
+      data: `User with id ${userId} successfully deleted`,
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/controllers/users/organizations-users.controller.ts:102]: I would return the deleted user here

File: apps/api/v2/src/modules/organizations/controllers/users/organizations-users.controller.ts [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+    @Param("orgId", ParseIntPipe) orgId: number,
+    @GetOrg() org: Team,
+    @Body() input: GetOrganizationsUsersInput
+  ): Promise<ApiResponse<GetOrganizationUsersOutput>> {
+    const users = await this.organizationsUsersService.getUsers(orgId, input.email);
+
+    return {
+      status: SUCCESS_STATUS,
+      data: {
+        users: users.map((user) =>
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/controllers/users/organizations-users.controller.ts:56]: I would return directly the list of users in data, no need for extra key

File: apps/api/v2/src/modules/organizations/controllers/users/organizations-users.controller.ts [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+export class OrganizationsUsersController {
+  constructor(private readonly organizationsUsersService: OrganizationsUsersService) {}
+
+  @Get()
+  @Roles("ORG_ADMIN")
+  async getOrganizationsUsers(
+    @Param("orgId", ParseIntPipe) orgId: number,
+    @GetOrg() org: Team,
+    @Body() input: GetOrganizationsUsersInput
+  ): Promise<ApiResponse<GetOrganizationUsersOutput>> {
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/controllers/users/organizations-users.controller.ts:50]: please copy the approach of apps/api/v2/src/modules/organizations/controllers/teams/organizations-teams.controller.ts to have a Response DTO that the swagger can generate doc automatically for 

check OrgTeamsOutputResponseDto on main

File: apps/api/v2/src/modules/organizations/controllers/users/organizations-users.e2e-spec.ts [hunk 20]
```diff
@@ -0,0 +1,279 @@
...
+      } catch (error) {}
+    });
+
+    afterAll(async () => {
+      // await membershipFixtures.delete(membership.id);
+      await Promise.all([
+        userRepositoryFixture.deleteByEmail(user.email),
+        userRepositoryFixture.deleteByEmail(nonMemberEmail),
+        ...orgMembersData.map((member) => userRepositoryFixture.deleteByEmail(member.email)),
+      ]);
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/controllers/users/organizations-users.e2e-spec.ts:274]: need to delete new-member@org.com as well

File: apps/api/v2/src/modules/organizations/controllers/users/organizations-users.e2e-spec.ts [hunk 21]
```diff
@@ -0,0 +1,279 @@
...
+    it("should not be able to update an org user", async () => {
+      return request(app.getHttpServer()).patch(`/v2/organizations/${org.id}/users/${user.id}`).expect(403);
+    });
+
+    it("should not be able to delete an org user", async () => {
+      return request(app.getHttpServer()).delete(`/v2/organizations/${org.id}/users/${user.id}`).expect(403);
+    });
+
+    afterAll(async () => {
+      await app.close();
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/controllers/users/organizations-users.e2e-spec.ts:88]: need to delete created user here

File: apps/api/v2/src/modules/organizations/controllers/users/organizations-users.controller.ts [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+      inviter.name ?? inviter.username ?? inviter.email
+    );
+    return {
+      status: SUCCESS_STATUS,
+      data: plainToInstance(GetOrganizationUserOutput, user, { strategy: "excludeAll" }),
+    };
+  }
+
+  @Patch("/:userId")
+  @Roles("ORG_ADMIN")
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/controllers/users/organizations-users.controller.ts:85]: @UseGuards(IsUserInOrg)

File: apps/api/v2/src/modules/organizations/controllers/users/organizations-users.controller.ts [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+  ): Promise<ApiResponse<GetOrganizationUserOutput>> {
+    const user = await this.organizationsUsersService.updateUser(orgId, userId, input);
+    return {
+      status: SUCCESS_STATUS,
+      data: plainToInstance(GetOrganizationUserOutput, user, { strategy: "excludeAll" }),
+    };
+  }
+
+  @Delete("/:userId")
+  @Roles("ORG_ADMIN")
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/controllers/users/organizations-users.controller.ts:100]: @UseGuards(IsUserInOrg)

File: apps/api/v2/src/modules/organizations/controllers/users/organizations-users.controller.ts [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+@UseGuards(IsOrgGuard)
+@DocsTags("Organizations Users")
+export class OrganizationsUsersController {
+  constructor(private readonly organizationsUsersService: OrganizationsUsersService) {}
+
+  @Get()
+  @Roles("ORG_ADMIN")
+  async getOrganizationsUsers(
+    @Param("orgId", ParseIntPipe) orgId: number,
+    @Body() input: GetOrganizationsUsersInput
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/controllers/users/organizations-users.controller.ts:51]: get endpoints don't have body, use ```@Query``` instead

**Review by ThyMinimalDev — APPROVED** [review]

**Pull request merged**